pkg/cloud/metadata.go

```diff
@@ -0,0 +1,101 @@
+package cloud
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+)
+
+// Metadata is info about the instance on which the driver is running
+type Metadata struct {
+	serviceInstanceID string
+	nodeInstanceId    string
+	region            string
+}
+
+// GetServiceInstanceId returns service instance id of the instance
+func (m *Metadata) GetServiceInstanceId() string {
+	return m.serviceInstanceID
+}
+
+// GetNodeInstanceId returns node instance id of the instance
+func (m *Metadata) GetNodeInstanceId() string {
+	return m.nodeInstanceId
+}
+
+// GetInstanceRegion returns region of the instance
+func (m *Metadata) GetInstanceRegion() string {
+	return m.region
+}
+
+type KubernetesAPIClient func() (kubernetes.Interface, error)
+
+// Get default kubernetes API client
+var DefaultKubernetesAPIClient = func() (kubernetes.Interface, error) {
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return clientset, nil
+}
+
+// Get New Metadata Service
+func NewMetadataService(k8sAPIClient KubernetesAPIClient) (MetadataService, error) {
+	klog.Infof("retrieving instance data from kubernetes api")
+	clientset, err := k8sAPIClient()
+	if err != nil {
+		klog.Warningf("error creating kubernetes api client: %v", err)
+	} else {
+		klog.Infof("kubernetes api is available")
+		return KubernetesAPIInstanceInfo(clientset)
+	}
+	return nil, fmt.Errorf("error getting instance data from ec2 metadata or kubernetes api")
+}
+
+// Get instance info from kubernetes API
+func KubernetesAPIInstanceInfo(clientset kubernetes.Interface) (*Metadata, error) {
+	nodeName := os.Getenv("CSI_NODE_NAME")
+	if nodeName == "" {
+		return nil, fmt.Errorf("CSI_NODE_NAME env var not set")
+	}
+	// get node with k8s API
+	node, err := clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting Node %v: %v", nodeName, err)
+	}
+
+	// Get node labels
+	labels := node.GetLabels()
+	keysList := []string{"serviceInstanceId", "nodeInstanceId", "region"}
+	instanceInfo := Metadata{}
+	for _, key := range keysList {
+		if val, ok := labels[key]; ok {
+			switch key {
+			case "serviceInstanceId":
+				instanceInfo.serviceInstanceID = val
+				break
+			case "nodeInstanceId":
+				instanceInfo.nodeInstanceId = val
```

Comment:
look into other providers how are they setting the provider-specific labels.

Reply:
this is the format I found from EBS CSI Driver,
```
topology.ebs.csi.aws.com/zone
```

Reply:
Few other formats from VMware CSI Driver:
```
        // Appplatform label that all vDPP PVCs must have.
	appplatformLabel = "appplatform.vmware.com/instance-id"
	// Opt out label present on pod to look for bound PVCs of all sibling replicas
	siblingReplicaCheckOptOutLabel = "psp.vmware.com/sibling-replica-check-opt-out"
	// Selected node annotation present on PVC
	pvcSelectedNode = "volume.kubernetes.io/selected-node"
	
```

Reply:
xref: https://github.com/kubernetes/kubernetes/blob/ae9ca48f01ddb03731e7903cfe91ef3db9ce8990/staging/src/k8s.io/api/core/v1/well_known_labels.go#L27

Reply:
I suggest we can use something like,

```
cloud.ibm.com/cloud-instance-id
cloud.ibm.con/pvm-instance-id
```